src/components/app/KeyboardShortcut.js

```diff
@@ -0,0 +1,270 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+
+import * as React from 'react';
+import { coerce } from '../../utils/flow';
+import classNames from 'classnames';
+
+import './KeyboardShortcut.css';
+
+type Props = {|
+  +wrapperClassName: string,
+  +children: React.Node,
+|};
+
+type State = {|
+  +isOpen: boolean,
+  +focusAfterClosed: HTMLElement | null,
+|};
+
+/**
+ * Display a list of shortcuts that overlays the screen.
+ */
+export class KeyboardShortcut extends React.PureComponent<Props, State> {
+  state = {
+    isOpen: false,
+    // This is a false positive due to how it's used, see the line:
+    //  `focusAfterClosed.focus()`
+    // eslint-disable-next-line react/no-unused-state
+    focusAfterClosed: null,
+  };
+
+  _focusArea = React.createRef<HTMLDivElement>();
+
+  componentDidMount() {
+    window.addEventListener('keydown', this._handleKeyPress);
+  }
+
+  _focus() {
+    requestAnimationFrame(() => {
+      // Only manipulate the DOM outside of the React setState update cycle, otherwise
+      // this triggers a React warning, plus we'll want the DOM to have been updated.
+      // However, this DOM update isn't future proof if React changes to a more
+      // asynchronous rendering method.
+      //
+      // "unstable_flushDiscreteUpdates: Cannot flush updates when React is already rendering."
+      const div = this._focusArea.current;
+      if (div) {
+        div.focus();
+      }
+    });
+  }
+
+  _open = (state: State): $Shape<State> => {
+    if (state.isOpen) {
+      // Do nothing.
+      return {};
+    }
+    const focusAfterClosed = document.activeElement;
+    this._trapFocus();
+    this._focus();
+    return { isOpen: true, focusAfterClosed };
+  };
+
+  _close = (state: State): $Shape<State> => {
+    const { focusAfterClosed, isOpen } = state;
+
+    if (!isOpen) {
+      // Do nothing.
+      return {};
+    }
+    this._untrapFocus();
+    if (focusAfterClosed) {
+      requestAnimationFrame(() => {
+        // Restore focus, but not during a React setState call, otherwise this triggers
+        // a React warning:
+        // "unstable_flushDiscreteUpdates: Cannot flush updates when React is already rendering."
+        focusAfterClosed.focus();
+      });
+    }
+
+    return { isOpen: false, focusAfterClosed: null };
+  };
+
+  _handleCloseClick = () => {
+    this.setState(this._close);
+  };
+
+  _handleKeyPress = (event: KeyboardEvent) => {
+    const target = coerce<EventTarget, HTMLElement>(event.target);
```

Comment:
I added it to a PR about a month or so back. It's nice because you can say what a type is currently, and what you are setting it to. I figure this will help be more precise in case the assumption for the coercion changes.